backend/cfg/cfg_polling.ml

```diff
@@ -0,0 +1,316 @@
+
+  let exception_ : domain -> (domain, error) result = fun dom -> Ok dom
+end
+
+let potentially_recursive_tailcall :
+    future_funcnames:String.Set.t ->
+    Cfg_with_layout.t ->
+    Polls_before_prtc_domain.t =
+ fun ~future_funcnames cfg_with_layout ->
+  let module Polls_before_prtc_transfer_ff = Polls_before_prtc_transfer (struct
+    let future_funcnames = future_funcnames
+  end) in
+  let module PTRCAnalysis =
+    Cfg_dataflow.Backward
+      (Polls_before_prtc_domain)
+      (Polls_before_prtc_transfer_ff)
+  in
+  let init : Polls_before_prtc_domain.t = Polls_before_prtc_domain.bot in
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  match PTRCAnalysis.run ~init ~map:PTRCAnalysis.Block cfg () with
+  | Ok res -> (
+    match Label.Tbl.find_opt res cfg.entry_label with
+    | None -> assert false
+    | Some res -> res)
+  | Aborted _ -> .
+  | Max_iterations_reached ->
+    Misc.fatal_error
+      "Cfg_polling.potentially_recursive_tailcall has been interrupted"
+
+let instr_cfg_with_layout :
+    Cfg_with_layout.t ->
+    block_needs_poll:PolledLoopsAnalysis.domain Label.Tbl.t ->
+    back_edges:Cfg_loop_infos.Edge.t list ->
+    bool =
+ fun cfg_with_layout ~block_needs_poll ~back_edges ->
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  List.fold_left back_edges ~init:false
+    ~f:(fun added_poll { Cfg_loop_infos.Edge.src; dst } ->
+      let needs_poll =
+        match Label.Tbl.find_opt block_needs_poll dst with
+        | None -> assert false
+        | Some Safe -> false
+        | Some Unsafe -> true
+      in
+      if needs_poll
+      then (
+        let after = Cfg.get_block_exn cfg src in
+        let cfg_infos = Regalloc_utils.collect_cfg_infos cfg_with_layout in
+        let next_instruction_id = ref (succ cfg_infos.max_instruction_id) in
+        let next_instruction_id () =
+          let res = !next_instruction_id in
+          incr next_instruction_id;
+          res
+        in
+        let poll =
+          (* CR xclerc for xclerc: double check the various fields. *)
+          { after.terminator with
+            Cfg.id = next_instruction_id ();
+            Cfg.desc = Cfg.Op Poll
+          }
+        in
+        (match
+           ( Label.Set.cardinal
+               (Cfg.successor_labels after ~normal:true ~exn:false),
+             after.exn )
+         with
+        | 1, None -> DLL.add_end after.body poll
+        | _ ->
+          let before = Some (Cfg.get_block_exn cfg dst) in
+          let instrs = DLL.of_list [poll] in
+          (* CR-soon xclerc: that kind of indicates the `insert_block` function
+             should be moved outside of "regalloc/" *)
+          let (_ : Cfg.basic_block list) =
+            Regalloc_utils.insert_block cfg_with_layout instrs ~after ~before
+              ~next_instruction_id
+          in
+          ());
+        true)
+      else added_poll)
+
+type polling_points = (polling_point * Debuginfo.t) list
+
+let add_poll_alloc_or_calls_basic :
```

Comment:
After the rebase, consider renaming this function to someting like `find_poll_or_alloc` and `find_calls` below.

Reply:
Changed the poll/alloc/call part for both
function, but kept the `add` prefix since
they are adding to an accumulator.